templates/aws-vpc.template.yaml

```diff
@@ -303,849 +345,911 @@ Parameters:
+                - "="
+                - !Ref "PublicSubnetTag3"
             Value: !Select
-              - '1'
+              - "1"
               - !Split
-                - '='
-                - !Ref 'PublicSubnetTag3'
-          - !Ref 'AWS::NoValue'
+                - "="
+                - !Ref "PublicSubnetTag3"
+          - !Ref "AWS::NoValue"
       MapPublicIpOnLaunch: true
   PrivateSubnet1ARouteTable:
     Condition: PrivateSubnetsCondition
     Type: AWS::EC2::RouteTable
     Properties:
-      VpcId: !Ref 'VPC'
+      VpcId: !Ref "VPC"
       Tags:
         - Key: Name
           Value: Private subnet 1A
         - Key: Network
           Value: Private
   PrivateSubnet1ARoute:
-    Condition: PrivateSubnetsCondition
+    Condition: PrivateSubnets&PublicSubnets&NatGatewaysCondition
     Type: AWS::EC2::Route
     Properties:
-      RouteTableId: !Ref 'PrivateSubnet1ARouteTable'
-      DestinationCidrBlock: '0.0.0.0/0'
-      NatGatewayId: !Ref 'NATGateway1'
+      RouteTableId: !Ref "PrivateSubnet1ARouteTable"
+      DestinationCidrBlock: "0.0.0.0/0"
+      NatGatewayId: !Ref "NATGateway1"
   PrivateSubnet1ARouteTableAssociation:
     Condition: PrivateSubnetsCondition
     Type: AWS::EC2::SubnetRouteTableAssociation
     Properties:
-      SubnetId: !Ref 'PrivateSubnet1A'
-      RouteTableId: !Ref 'PrivateSubnet1ARouteTable'
+      SubnetId: !Ref "PrivateSubnet1A"
+      RouteTableId: !Ref "PrivateSubnet1ARouteTable"
   PrivateSubnet2ARouteTable:
     Condition: PrivateSubnetsCondition
     Type: AWS::EC2::RouteTable
     Properties:
-      VpcId: !Ref 'VPC'
+      VpcId: !Ref "VPC"
       Tags:
         - Key: Name
           Value: Private subnet 2A
         - Key: Network
           Value: Private
   PrivateSubnet2ARoute:
-    Condition: PrivateSubnetsCondition
+    Condition: PrivateSubnets&PublicSubnets&NatGatewaysCondition
     Type: AWS::EC2::Route
     Properties:
-      RouteTableId: !Ref 'PrivateSubnet2ARouteTable'
-      DestinationCidrBlock: '0.0.0.0/0'
-      NatGatewayId: !Ref 'NATGateway2'
+      RouteTableId: !Ref "PrivateSubnet2ARouteTable"
+      DestinationCidrBlock: "0.0.0.0/0"
+      NatGatewayId: !Ref "NATGateway2"
   PrivateSubnet2ARouteTableAssociation:
     Condition: PrivateSubnetsCondition
     Type: AWS::EC2::SubnetRouteTableAssociation
     Properties:
-      SubnetId: !Ref 'PrivateSubnet2A'
-      RouteTableId: !Ref 'PrivateSubnet2ARouteTable'
+      SubnetId: !Ref "PrivateSubnet2A"
+      RouteTableId: !Ref "PrivateSubnet2ARouteTable"
   PrivateSubnet3ARouteTable:
     Condition: PrivateSubnets&3AZCondition
     Type: AWS::EC2::RouteTable
     Properties:
-      VpcId: !Ref 'VPC'
+      VpcId: !Ref "VPC"
       Tags:
         - Key: Name
           Value: Private subnet 3A
         - Key: Network
           Value: Private
   PrivateSubnet3ARoute:
-    Condition: PrivateSubnets&3AZCondition
+    Condition: PrivateSubnets&PublicSubnets&NatGateways&3AZCondition
```

Comment:
Instead of this long condition, you can just create Natgateway&3AZCondition, and use it here. As private and public both subnet will be anyways true if natgateway is true.